### PR TITLE
[Merged by Bors] - feat(Polynomial/AlgebraMap): add `mapAlgHom`

### PR DESCRIPTION
**Pull request by quangvdao**
Add `Polynomial.mapAlgHom` as an algebra homomorphism between two polynomial rings, extending `Polynomial.mapRingHom`. Provide supporting theorems for `mapAlgHom`.

Moved `Polynomial.eval₂AlgHom'` to earlier in `AlgebraMap.lean` , so that I can provide a theorem linking `mapAlgHom` and `eval₂AlgHom'`, without having to put `mapAlgHom` stuff after `aeval`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Polynomial/AlgebraMap.lean [hunk 1]
```diff
@@ -173,6 +165,56 @@ def eval₂AlgHom' (f : A →ₐ[R] B) (b : B) (hf : ∀ a, Commute (f a) b) : A
...
+theorem mapAlgHom_comp (C : Type z) [Semiring C] [Algebra R C] (f : B →ₐ[R] C) (g : A →ₐ[R] B) :
+    (mapAlgHom f).comp (mapAlgHom g) = mapAlgHom (f.comp g) := by
+  apply AlgHom.ext
+  intro x
+  simp [AlgHom.comp_algebraMap, map_map]
+  congr
+
+/-- If `A` and `B` are isomorphic as `R`-algebras, then so are their polynomial rings -/
+def mapAlgEquiv (f : A ≃ₐ[R] B) : Polynomial A ≃ₐ[R] Polynomial B :=
+  AlgEquiv.ofAlgHom (mapAlgHom f.toAlgHom) (mapAlgHom f.symm.toAlgHom) (by simp) (by simp)
```
Comment (semorrison) [inline Mathlib/Algebra/Polynomial/AlgebraMap.lean:200]: This needs analogues of the 4 theorems above.

Comment (acmepjz) [reply Mathlib/Algebra/Polynomial/AlgebraMap.lean:200]: Do we have a stronger result `Polynomial A ≃ₐ[Polynomial R] Polynomial B`?

Comment (quangvdao) [reply Mathlib/Algebra/Polynomial/AlgebraMap.lean:200]: I can add in those results. I am not sure how to name them though. `mapAlgHom'` and `mapAlgEquiv'` is an option, but they are not very descriptive.

Comment (erdOne) [reply Mathlib/Algebra/Polynomial/AlgebraMap.lean:206]: I think they could replace `mapAlgHom`. If you need the `R`-alghom version, use `AlgHom.restrictScalars`.

**Comment by acmepjz** [comment]
Sorry but I think I made a mistake. The instance `Algebra (Polynomial R) (Polynomial A)` cannot be inferred automatically. And that should not be added to mathlib in my opinion, since potentially the intermediate X is not necessary map to X.

**Comment by quangvdao** [comment]
@acmepjz: yeah, I agree. So let's just go ahead and add `mapAlgHom`, `mapAlgEquiv` the way I currently write it?

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**